src/app/server-cluster/OptionalAttributes.h

```diff
@@ -0,0 +1,139 @@
+/*
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/DataModelTypes.h>
+
+namespace chip {
+namespace app {
+namespace Internal {
+
+// Compile-time helper to check if an ID is in a list of IDs.
+template <AttributeId T, AttributeId... Ts>
+struct IsOneOf;
+
+template <AttributeId T, AttributeId Head, AttributeId... Tail>
+struct IsOneOf<T, Head, Tail...>
+{
+    static constexpr bool value = (T == Head) || IsOneOf<T, Tail...>::value;
+};
+
+template <AttributeId T>
+struct IsOneOf<T>
+{
+    static constexpr bool value = false;
+};
+
+} // namespace Internal
+
+/// It is very common that a class has optional attributes. Such optional attributes
+/// need checking for and also affect what attributes are being returned by
+/// server cluster implementations
+///
+/// This class implements the common case where attribute IDs can fit
+/// within a 32-bit bitset as most attributes start with low IDs and are incremented
+/// by one.
```

Comment:
This is obviously going to fail for some clusters (On/Off, Level Control, Door Lock, etc).  Are those clusters just going to not be able to use AttributeListBuilder?  Or will we come up with something at that point?

I'm a little worried that we'll end up needing two separate systems for this anyway, and some common small devices (light bulbs) will need to pay the codesize cost for both, because we know for sure this setup will not work for some of the clusters those devices need.

Reply:
I did not expect basic clusters to use high IDs ... that is indeed not ideal. Looking to see if we can do better in this case ... OnOff and LevelControl are indeed something we want to support on very small devices.

Reply:
we also seem to have a few clusters with a lot of attributes - ThreadNetworkDiagnostics has 70, Thermostat has 69 (and then beyond 31 we have ColorControl with 57, CameraAvStreamManagement with 46 and DoorLock with 50). Granted not all are optional, however they are guaranteed to be over the bitset even if we use a 64-bit bitset.


Having said that however, this is saving flash today for all clusters not falling in the above category, so I believe even if we do not come up with a "common method" it may still be worth it because a lot of clusters can actually use this (but not OnOff,Level,DoorLock,Color and so on).

I am afraid that any generalization would involve some form of vtable or mapping storage which will have some additional cost. We may want to figure that out in stages.